heron/tools/cli/src/python/response.py

```diff
@@ -68,33 +70,55 @@ def add_context(self, err_context, succ_context=None):
     self.err_context = err_context
     self.succ_context = succ_context
 
+def do_log(log, msg):
+  if msg:
+    log(msg)
+
+def log_context(resp):
+  status = resp.status
+  if status == Status.Ok or status == Status.DryRun:
+    do_log(Log.info, resp.succ_context)
+  elif status == Status.HeronError:
+    do_log(Log.error, resp.err_context)
+  elif status == Status.InvocationError:
+    pass
+  else:
+    raise RuntimeError("Unknown status type of value %d", status)
+
+# # pylint: disable=too-many-branches,unreachable
 def render(resp):
-  def do_log(log, msg):
-    if msg:
-      log(msg)
+  """ Render output of the process until it exits
+
+  :param resp: Response
+  :return:
+  """
   if isinstance(resp, list):
     for r in resp:
       render(r)
   elif isinstance(resp, Response):
-    if resp.status == Status.Ok:
-      do_log(Log.info, resp.succ_context)
-      do_log(Log.info, resp.msg)
-      do_log(Log.debug, resp.detailed_msg)
-    elif resp.status == Status.HeronError:
-      do_log(Log.error, resp.err_context)
-      do_log(Log.error, resp.msg)
-      do_log(Log.debug, resp.detailed_msg)
-    # If status code is InvocationError, invocation of shelled-out program fails. The error
-    # message will be in stderr, so we log.error detailed message(stderr) only
-    elif resp.status == Status.InvocationError:
-      do_log(Log.error, resp.detailed_msg)
-    elif resp.status == Status.DryRun:
-      do_log(Log.info, resp.succ_context)
-      # No need to prefix [INFO] here. We want to display dry-run response in a clean way
-      print resp.msg
-      do_log(Log.debug, resp.detailed_msg)
+    if resp.proc:
+      proc = resp.proc
+      # use while loop here because of a bug of Python2.7
+      while True:
+        stderr_line = proc.stderr.readline()
```

Comment:
In this method you're reading stderr, polling if None and then readying stdout, then polling if None. Are their chances for raceconditions here due to state changing on proc between calls? Also are there edge conditions possible when both stderr and stdout are set? I don't know the specifics of the proc API in this sense to say for sure.